purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt

```diff
@@ -0,0 +1,394 @@
+package com.revenuecat.purchases
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.revenuecat.purchases.common.sha1
+import com.revenuecat.purchases.factories.StoreProductFactory
+import com.revenuecat.purchases.factories.StoreTransactionFactory
+import com.revenuecat.purchases.helpers.mockQueryProductDetails
+import com.revenuecat.purchases.interfaces.UpdatedCustomerInfoListener
+import com.revenuecat.purchases.models.StoreTransaction
+import io.mockk.every
+import io.mockk.verify
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.fail
+import org.junit.After
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.time.Duration
+
+@Suppress("TooManyFunctions")
+@RunWith(AndroidJUnit4::class)
+class OfflineEntitlementsIntegrationTest : BasePurchasesIntegrationTest() {
+
+    private val initialActiveTransaction = StoreTransactionFactory.createStoreTransaction(
+        skus = listOf(Constants.productIdToPurchase),
+        purchaseToken = Constants.googlePurchaseToken
+    )
+    private val initialActivePurchases = mapOf(
+        initialActiveTransaction.purchaseToken.sha1() to initialActiveTransaction
+    )
+
+    @After
+    fun tearDown() {
+        tearDownTest()
+    }
+
+    @Test
+    fun doesNotEnterOfflineEntitlementsModeIfNoProductEntitlementMappingAvailable() {
+        ensureBlockFinishes { latch ->
+            setupTest(
+                initialActivePurchases = initialActivePurchases,
+                forceServerErrors = true
+            ) {
+                Purchases.sharedInstance.getCustomerInfoWith(
+                    onError = {
+                        latch.countDown()
+                        assertThat(it.code).isEqualTo(PurchasesErrorCode.UnknownBackendError)
+                    },
+                    onSuccess = {
+                        latch.countDown()
+                        fail("Expected error")
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfNoCachedCustomerInfoAndCustomerInfoRequestReturns500() {
+        ensureBlockFinishes { latch ->
+            setupTestWaitingForInitialRequests(
```

Comment:
To simplify these, I'd recommend combining tests that share the same set up (`setupTest` / `setupTestWaitingForInitialRequests`) into separate classes, so this can be done in the `@Before`. 
I try to follow that pattern to make the tests themselves easier to follow so they only focus on what they have to do once the SDK is already initialized (in different ways).
What do you think?

Reply:
Well, with async/coroutines that would be much simpler, but we need to wait to run the test until the callback is received here, so we can't move it to the `@Before` easily... I guess we could create a latch to wait for that in the `@Before`. We also need the `activity` in some of these, though I guess we could cache it ourselves... Will try that to try to simplify these tests.